lambdas/backend-api/src/templates/domain/template/template-schema.ts

```diff
@@ -0,0 +1,79 @@
+import { z } from 'zod';
+import {
+  CreateTemplate,
+  TemplateStatus,
+  TemplateType,
+  UpdateTemplate,
+} from 'nhs-notify-backend-client';
+import { schemaFor } from '@backend-api/utils/schema-for';
+import {
+  MAX_SMS_CHARACTER_LENGTH,
+  MAX_EMAIL_CHARACTER_LENGTH,
+  MAX_NHS_APP_CHARACTER_LENGTH,
+  MAX_LETTER_CHARACTER_LENGTH,
+  NHS_APP_DISALLOWED_CHARACTERS,
+} from './constants';
+
+const $Template = schemaFor<CreateTemplate>()(
+  z.object({
+    templateType: z.nativeEnum(TemplateType),
+    name: z.string().min(1),
+    message: z.string(),
+  })
+);
+
+export const $SMSTemplate = schemaFor<CreateTemplate>()(
+  $Template.extend({
+    templateType: z.literal(TemplateType.SMS),
+    message: z.string().min(1).max(MAX_SMS_CHARACTER_LENGTH),
+  })
+);
+
+export const $NhsAppTemplate = schemaFor<CreateTemplate>()(
+  $Template.extend({
+    templateType: z.literal(TemplateType.NHS_APP),
+    message: z
+      .string()
+      .min(1)
+      .max(MAX_NHS_APP_CHARACTER_LENGTH)
+      // eslint-disable-next-line security/detect-non-literal-regexp
+      .refine((s) => !new RegExp(NHS_APP_DISALLOWED_CHARACTERS, 'gi').test(s), {
```

Comment:
if you use regex like `/[a-Z]/gi.test('pickle')`  as opposed to n`ew Regex('[a-Z]', 'gi).test('pickle')` .

The literal regex` /[a-Z]/` will sometimes pass/fail if you call it multiple times.

It's a bug in JS where the matching index is sometimes not reset and thus the wonky behaviour. Newing up a Regex on each call will always reset the matching indexes to 0 and will consistently work as expected.